### PR TITLE
Adding state to to WaiterError exception

### PR DESCRIPTION
**Pull request by zdutta**
Improving WaiterError exception message with failure state. Addressing https://github.com/aws/aws-cli/issues/2918

### CODE REVIEW

File: botocore/waiter.py [hunk 1]
```diff
@@ -309,23 +324,35 @@ def wait(self, **kwargs):
...
+                        ),
+                        last_response=response,
                     )
             if current_state == 'success':
                 logger.debug("Waiting complete, waiter matched the "
                              "success state.")
                 return
             if current_state == 'failure':
+                reason = 'Waiter encountered a terminal failure state: %s' % (
+                        acceptor.explanation
```
Comment (nateprewitt) [inline botocore/waiter.py:339]: This will potentially be None if we implement a new acceptor later and blow up with a `ValueError` on formatting. Should be set up a safer default in the event of no match?

File: tests/unit/test_waiters.py [hunk 2]
```diff
@@ -392,6 +392,53 @@ def test_unspecified_errors_propagate_error_code(self):
...
+            ([{'state': 'failure', 'matcher': 'status', 'expected': 404}],
+             [{'ResponseMetadata': {'HTTPStatusCode': 404}}],
+             '404'),
+            ([{'state': 'failure', 'matcher': 'error', 'expected': 'FailError'}],
+             [{'Error': {'Code': 'FailError', 'Message': 'foo'}}],
+             'FailError'),
+            ([{'state': 'retry', 'matcher': 'error', 'expected': 'RetryMe'}],
+             [{'Success': False}]*4,
+             'Max attempts exceeded'),
+            ([{'state': 'retry', 'matcher': 'error', 'expected': 'RetryMe'}],
```
Comment (nateprewitt) [inline tests/unit/test_waiters.py:430]: Can we keep this test consistent with the one we deleted. That way we're sure we're not losing any edge cases.

**Review by nateprewitt — CHANGES_REQUESTED** [review]

File: botocore/waiter.py (hunk 1 above)
Comment (joguSD) [reply botocore/waiter.py:339]: @nateprewitt In practice it's probably unlikely. If an unknown waiter type is encountered you can't actually instantiate the `AcceptorConfig` object anyways. Though I suppose a default message long the lines of `'No explanation for unknown waiter type "%s"'`.

**Review by nateprewitt — APPROVED** [review]
